### PR TITLE
Replace worker_hash calls with partkey IS NOT NULL filters

### LINKED ISSUE #1671: Citus adds extra filter on INSERT / SELECT queries which messes PG planner estimates
While running INSERT / SELECT queries, Citus adds filters on worker node queries like:

```
((worker_hash(distribution column) >= '-2147483648'::integer) AND (worker_hash(distribution column) <= '-2013265921'::integer) 
```

This leads to the planner having very wrong estimates. An example is below.

```
->  Seq Scan on table  (cost=0.00..2742400.71 rows=1430 width=66) (actual
 time=43751.529..119194.216 rows=16207353 loops=1)
                                       Filter: ((worker_hash(column) >= '-2147483648'::integer) AND
 (worker_hash(column) <= '-2013265921'::integer) AND (to_timestamp(("timestamp")::double precision)
 >= '2017-09-19 00:00:00+00'::timestamp with time zone) AND (to_timestamp(("timestamp")::double
 precision) < '2017-09-20 00:00:00+00'::timestamp with time zone))
```

This can lead to worse query plans on each worker node shard and hence worse query performance.

### PR DESCRIPTION
**Pull request by hanefi**
DESCRIPTION: Improves cost estimation of INSERT SELECT plans

In https://github.com/citusdata/citus/issues/1671#issuecomment-618828549 @onderkalaci said
> As far as I remember, we had to add worker_hash filters because the way we detect unsupported outer joins was failing to detect all cases when reference tables involved (it was the initial implementation of INSERT .. SELECT pushdown, we've fixed that already). That's why, for certain types of OUTER JOINs, it was possible to INSERT NULL values to distribution key on the target table. To filter out those, we added worker_hash calls.

I went over all codepaths that added `worker_hash` calls to filters in subqueries, and found out that all of them are distributed tables, and they are only there to check for the existence of partition column values.

Fixes #1671

### CODE REVIEW

File: src/test/regress/expected/isolation_append_copy_vs_all.out [hunk 1]
```diff
@@ -102,7 +102,7 @@ step s1-commit: COMMIT;
 step s1-select-count: SELECT COUNT(*) FROM append_copy;
 count
 
-10
+15
```
Comment (hanefi) [inline src/test/regress/expected/isolation_append_copy_vs_all.out:105]: This change is quite strange. 15 looks like the correct result to me, but I need to make sure.

@onderkalaci do you think this change is expected here?

File: src/test/regress/expected/isolation_range_copy_vs_all.out [hunk 2]
```diff
@@ -102,7 +102,7 @@ step s1-commit: COMMIT;
 step s1-select-count: SELECT COUNT(*) FROM range_copy;
 count
 
-10
+15
```
Comment (hanefi) [inline src/test/regress/expected/isolation_range_copy_vs_all.out:105]: Investigating why this change happened. Similar to https://github.com/citusdata/citus/pull/4207/files#r497480623

File: src/test/regress/expected/isolation_select_vs_all.out [hunk 3]
```diff
@@ -308,7 +308,7 @@ step s1-commit: COMMIT;
 step s1-select-count: SELECT COUNT(*) FROM select_append;
 count
 
-5
+10
```
Comment (hanefi) [inline src/test/regress/expected/isolation_select_vs_all.out:311]: Investigating why this change happened. Similar to https://github.com/citusdata/citus/pull/4207/files#r497480623

**Comment by onderkalaci** [comment]
To be fair, it is only a theory that converting `worker_hash` to `IS NOT NULL` would yield better plans on the shards. Could you maybe try to verify that? So, create a table, load some large amounts of data and try to see if the the query plans on the shards are betters?

File: src/test/regress/expected/cte_inline_0.out [hunk 4]
```diff
@@ -761,10 +761,10 @@ WITH fist_table_cte AS
...
+DEBUG:  distributed statement: INSERT INTO cte_inline.test_table_1960000 AS citus_table_alias (key, value) SELECT key, value FROM (SELECT test_table.key, test_table.value, test_table.other_value FROM cte_inline.test_table_1960000 test_table) fist_table_cte WHERE (key IS NOT NULL)
+DEBUG:  distributed statement: INSERT INTO cte_inline.test_table_1960001 AS citus_table_alias (key, value) SELECT key, value FROM (SELECT test_table.key, test_table.value, test_table.other_value FROM cte_inline.test_table_1960001 test_table) fist_table_cte WHERE (key IS NOT NULL)
+DEBUG:  distributed statement: INSERT INTO cte_inline.test_table_1960002 AS citus_table_alias (key, value) SELECT key, value FROM (SELECT test_table.key, test_table.value, test_table.other_value FROM cte_inline.test_table_1960002 test_table) fist_table_cte WHERE (key IS NOT NULL)
+DEBUG:  distributed statement: INSERT INTO cte_inline.test_table_1960003 AS citus_table_alias (key, value) SELECT key, value FROM (SELECT test_table.key, test_table.value, test_table.other_value FROM cte_inline.test_table_1960003 test_table) fist_table_cte WHERE (key IS NOT NULL)
```
Comment (marcocitus) [inline src/test/regress/expected/cte_inline_0.out:767]: this seems kind of ok, not worse than before, but it does make wonder whether we should be auto-creating not null constraints on the distribution column instead of adding this filter

**Comment by hanefi** [comment]
> To be fair, it is only a theory that converting `worker_hash` to `IS NOT NULL` would yield better plans on the shards. Could you maybe try to verify that? So, create a table, load some large amounts of data and try to see if the the query plans on the shards are betters?

I created a test setup with 3 consecutive EXPLAIN INSERT SELECT runs. The row estimates on master branch were roughly 1/200 of the actual number of rows whereas on this branch their difference was less than 1%.

Overall I think this is the improvement that we aimed for.

### Sequential scan nodes on master:

Seq Scan on public.t1_102053 t1  (cost=0.00..7632.54 rows=**1562** width=4) (actual time=0.191..1150.167 rows=**312477** loops=1)

Seq Scan on public.t1_102047 t1  (cost=0.00..15289.48 rows=**3130** width=4) (actual time=0.030..2250.564 rows=**626150** loops=1)

Seq Scan on public.t1_102053 t1  (cost=0.00..30524.64 rows=**6248** width=4) (actual time=0.027..3723.725 rows=**1249908** loops=1)

------

### Sequential scan nodes on feature branch:

Seq Scan on public.t1_102051 t1  (cost=0.00..4507.86 rows=**312486** width=4) (actual time=0.046..258.888 rows=**312486** loops=1)

Seq Scan on public.t1_102061 t1  (cost=0.00..9041.62 rows=**626762** width=4) (actual time=0.033..607.932 rows=**626762** loops=1)

Seq Scan on public.t1_102047 t1  (cost=0.00..18060.48 rows=**1251848** width=4) (actual time=2.533..1199.567 rows=**1252300** loops=1)

File: src/test/regress/expected/cte_inline_0.out (hunk 4 above)
Comment (hanefi) [reply src/test/regress/expected/cte_inline_0.out:767]: Even if we have a not null constraint on the distribution column values, a FULL JOIN can still yield NULL values for these columns. I think we should still keep some kind of filter if we wish to keep the current behaviour.

```sql
CREATE TABLE a(x int not null);
CREATE TABLE b(x int not null);
INSERT INTO a VALUES (1);
INSERT INTO b VALUES (2);
SELECT a.x, b.x FROM a FULL JOIN b USING (x);
┌────────┬────────┐
│   x    │   x    │
├────────┼────────┤
│      1 │ [NULL] │
│ [NULL] │      2 │
└────────┴────────┘
```

If we were to add not null constraints and remove these filters in question, Citus will behave similar to Vanilla PG. The following query fails on vanilla PG, but succeeds on Citus ( if a and b are distributed ) by silently ignoring null values.

```sql
INSERT INTO a SELECT a.x FROM a FULL JOIN b USING (x);
ERROR:  23502: null value in column "x" violates not-null constraint
DETAIL:  Failing row contains (null).
SCHEMA NAME:  public
TABLE NAME:  a
COLUMN NAME:  x
```

**Comment by onderkalaci** [comment]
> I created a test setup with 3 consecutive EXPLAIN INSERT SELECT runs. The row estimates on master branch were roughly 1/200 of the actual number of rows whereas on this branch their difference was less than 1%.

Overall I think this is the improvement that we aimed for.

can you share the query for reference?

**Comment by hanefi** [comment]
```sql
CREATE TABLE t1 (a int);
CREATE TABLE t2 (a int);
SELECT create_distributed_table('t1', 'a');
SELECT create_distributed_table('t2', 'a');

INSERT INTO t1 SELECT x FROM generate_series(1,10000000) x;
INSERT INTO t2 SELECT x FROM generate_series(1,100) x;
```

Then on master and feature branch I ran the following:

```sql
BEGIN;
EXPLAIN ANALYZE VERBOSE INSERT INTO t1 SELECT t1.a FROM t1 FULL JOIN t2 USING (a) RETURNING * ;
EXPLAIN ANALYZE VERBOSE INSERT INTO t1 SELECT t1.a FROM t1 FULL JOIN t2 USING (a) RETURNING * ;
EXPLAIN ANALYZE VERBOSE INSERT INTO t1 SELECT t1.a FROM t1 FULL JOIN t2 USING (a) RETURNING * ;
ROLLBACK;
```

I have 6 plans (3 on master, 3 on my branch) with t1 having 10M, 20M, 40M rows. I can share them if you want.

**Comment by hanefi** [comment]
I should also note that I ran VACUUM before collecting the plans.

```sql
VACUUM ANALYSE;
VACUUM ANALYSE t1;
VACUUM ANALYSE t2;
```

**Review by marcocitus — APPROVED** [review]
This all seems fine. 

One thing we may want to change in the future is to stop pushing down INSERT..SELECT commands which use the distribution column on the inside of an outer join in the target list. That will break those queries, so it can only be done in a major release.
